src/main/java/jenkins/plugins/openstack/compute/SectionDisabled.java

```diff
@@ -0,0 +1,277 @@
+     * @throws DisabledException is {@link #isDisabled()} is true.
+     */
+    @Restricted(NoExternalUse.class)
+    public void throwIfDisabled(@Nullable String exceptionMsgPrefix) {
+        final String disabledMsg = toString();
+        if (NOT_DISABLED.equals(disabledMsg)) {
+            return;
+        }
+        final String msg = Util.fixEmpty(exceptionMsgPrefix) == null ? disabledMsg : exceptionMsgPrefix + disabledMsg;
+        throw new DisabledException(msg, exceptionWhenDisabledBySystem);
+    }
+
+    /**
+     * An {@link IllegalStateException} thrown by
+     * {@link SectionDisabled#throwIfDisabled(String)}.
+     */
+    public static class DisabledException extends IllegalStateException {
+        private static final long serialVersionUID = 1L;
+
+        @Restricted(NoExternalUse.class)
+        public DisabledException(String msg, Throwable cause) {
+            super(msg, cause);
+        }
+    }
+
+    @Extension
+    public static class DescriptorImpl extends Descriptor<SectionDisabled> {
+        public FormValidation doCheckEnabledByChoice(@QueryParameter boolean enabledByChoice,
+                @QueryParameter boolean disabledBySystem, @QueryParameter String whenDisabledBySystemString,
+                @QueryParameter String whenReEnableBySystemString, @QueryParameter String reasonWhyDisabledBySystem,
+                @QueryParameter String exceptionWhenDisabledBySystemString) {
+            if (!enabledByChoice) {
+                return FormValidation.warning("Note: Disabled.");
+            }
+            if (disabledBySystem) {
+                final String reason = Util.fixNull(reasonWhyDisabledBySystem);
+                final String disabledAgo = Util.fixNull(whenDisabledBySystemString);
+                final String enableWhen = Util.fixNull(whenReEnableBySystemString);
+                final String exception = Util.fixNull(exceptionWhenDisabledBySystemString);
+                if (!reason.isEmpty() && !disabledAgo.isEmpty() && !enableWhen.isEmpty()) {
+                    final StringBuilder html = new StringBuilder();
+                    html.append("Note: Disabled ");
+                    html.append(Util.escape(disabledAgo));
+                    html.append(" ago due to error.");
+                    html.append("  Will re-enable in ");
+                    html.append(Util.escape(enableWhen));
+                    html.append(".");
+                    html.append("<br/>Reason: ");
+                    html.append(Util.escape(reason));
+                    if (!exception.isEmpty()) {
+                        html.append(" <a href='#' class='showDetails'>");
+                        html.append(Messages.ShowDetails());
+                        html.append("</a><pre style='display:none'>");
+                        html.append(Util.escape(exception));
+                        html.append("</pre>");
+                    }
+                    return FormValidation.warningWithMarkup(html.toString());
```

Comment:
Is there a reason for not using `FormValidation#warning(Throwable, String)`

Reply:
I can't remember 100%, but I suspect that `#warning` probably didn't look nice enough when I originally wrote this code for the docker-plugin.
...or it might just be ignorance - maybe I didn't spot that that method existed.
Either is possible, but I suspect the former is more likely.